File: pipeline/kite9-pipeline-common/src/main/java/com/kite9/k9server/command/xml/insert/InsertXML.java
package com.kite9.k9server.command.xml.insert;

import org.w3c.dom.Element;

import com.kite9.k9server.adl.holder.pipeline.ADLDom;
import com.kite9.k9server.command.CommandException;

/**
 * As well as inserting, you can also use this to "surround" an element.  Not sure if
 * we'll use that yet, as you can achive the same result with move.
 * 
 * @author robmoffat
 *
 */
public class InsertXML extends AbstractInsertCommand {

	public String base64Element;

	@Override
	public Mismatch applyCommand(ADLDom in) throws CommandException {
		checkProperties();
		return doInsert(in);
	}

	@Override
	public Mismatch undoCommand(ADLDom in) throws CommandException {
		checkProperties();
		return doDelete(in);
	}

	@Override
	protected Element getContents(ADLDom in) {
		Element out = copyWithoutContainedIds(decodeElement(base64Element, in));
		if (newId != null) {
			replaceIds(out, newId);
		}
		return out;
	}

	@Override
	protected void checkProperties() {
		super.checkProperties();
		ensureNotNull("base64Element", fragmentId);
	}


}

File: b/src/jvmMain/kotlin/com/kite9/pipeline/command/xml/insert/InsertXML.kt
package com.kite9.pipeline.command.xml.insert

import com.kite9.pipeline.adl.holder.pipeline.ADLDom
import com.kite9.pipeline.command.Command.Mismatch
import com.kite9.pipeline.command.CommandContext
import org.w3c.dom.Element

/**
 * As well as inserting, you can also use this to "surround" an element.  Not sure if
 * we'll use that yet, as you can achive the same result with move.
 *
 * @author robmoffat
 */
class InsertXML : AbstractInsertCommand() {

    @JvmField
    var base64Element: String? = null

    override fun applyCommand(d: ADLDom, ctx: CommandContext): Mismatch? {
        checkProperties()
        return doInsert(d, ctx)
    }

    override fun undoCommand(d: ADLDom, ctx: CommandContext): Mismatch? {
        checkProperties()
        return doDelete(d, ctx)
    }

    override fun getContents(d: ADLDom, ctx: CommandContext): Element? {
        val out = copyWithoutContainedIds(ctx.decodeElement(base64Element, d))
        if (newId != null) {
            replaceIds(out!!, newId!!)
        }
        return out
    }

    override fun checkProperties() {
        super.checkProperties()
        ensureNotNull("base64Element", fragmentId)
    }
}